azimuth_schedule_operator/openstack/openstack.py

```diff
@@ -0,0 +1,227 @@
+import asyncio
+import contextlib
+import urllib.parse
+
+import httpx
+
+from easykube import rest
+
+
+class UnsupportedAuthenticationError(Exception):
+    """Raised when an unsupported authentication method is used."""
+
+    def __init__(self, auth_type):
+        super().__init__(f"unsupported authentication type: {auth_type}")
+
+
+class Auth(httpx.Auth):
+    """Authenticator class for OpenStack connections."""
+
+    def __init__(
+        self, auth_url, application_credential_id, application_credential_secret
+    ):
+        self.url = auth_url
+        self._application_credential_id = application_credential_id
+        self._application_credential_secret = application_credential_secret
+        self._token = None
+        self._user_id = None
+        self._lock = asyncio.Lock()
+
+    @contextlib.asynccontextmanager
+    async def _refresh_token(self):
+        """
+        Context manager to ensure only one request at a time triggers a token refresh.
+        """
+        token = self._token
+        async with self._lock:
+            # Only yield to the wrapped block if the token has not changed
+            # in the time it took to acquire the lock
+            if token == self._token:
+                yield
+
+    def _build_token_request(self):
+        return httpx.Request(
+            "POST",
+            f"{self.url}/v3/auth/tokens",
+            json={
+                "auth": {
+                    "identity": {
+                        "methods": ["application_credential"],
+                        "application_credential": {
+                            "id": self._application_credential_id,
+                            "secret": self._application_credential_secret,
+                        },
+                    },
+                },
+            },
+        )
+
+    def _handle_token_response(self, response):
+        response.raise_for_status()
+        self._token = response.headers["X-Subject-Token"]
+        self._user_id = response.json()["token"]["user"]["id"]
+
+    async def async_auth_flow(self, request):
+        if self._token is None:
+            async with self._refresh_token():
+                response = yield self._build_token_request()
+                await response.aread()
+                self._handle_token_response(response)
+        request.headers["X-Auth-Token"] = self._token
+        # TODO(johngarbutt): this is needed for blazar
+        request.headers["Content-Type"] = "application/json"
+        # TODO(johngarbutt): this is needed for nova flavor extra spec info
+        request.headers["X-OpenStack-Nova-API-Version"] = "2.61"
```

Comment:
@mkjpryor to make this work I needed these two hacks over and above the library I have cut and paste into here. I didn't work hard on how to add this properly.